src/esp/gfx_batch/Hbao.cpp

```diff
@@ -0,0 +1,1088 @@
+
+  /* "init misc" */
+  {
+    std::mt19937 rmt;
+
+    float numDir = 8;  // keep in sync to glsl
+    for (std::size_t i = 0; i < HbaoRandomNumElements * MaxSamples; ++i) {
+      // TODO use better random gen method
+      float rand1 = Mn::Float(rmt()) / 4294967296.0f;
+      float rand2 = Mn::Float(rmt()) / 4294967296.0f;
+
+      // Use random rotation angles in [0,2PI/NUM_DIRECTIONS)
+      Mn::Rad angle{Mn::Constants::tau() * rand1 / numDir};
+
+      state_->random[i] = {Mn::Math::cos(angle), Mn::Math::sin(angle), rand2,
+                           0.0f};
+    }
+
+    /* Pack those to 16-bit and upload to a texture */
+    Mn::Vector4s hbaoRandomShort[HbaoRandomNumElements * MaxSamples];
+    Mn::Math::packInto(Cr::Containers::arrayCast<2, Mn::Float>(
+                           Cr::Containers::stridedArrayView(state_->random)),
+                       Cr::Containers::arrayCast<2, Mn::Short>(
+                           Cr::Containers::stridedArrayView(hbaoRandomShort)));
+
+    constexpr Mn::Vector3i size{Mn::Vector2i{HbaoRandomSize}, MaxSamples};
+
+    state_->hbaoRandom.setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setStorage(1, Mn::GL::TextureFormat::RGBA16Snorm, size)
+        .setSubImage(0, {},
+                     Mn::ImageView3D{Mn::PixelFormat::RGBA16Snorm, size,
+                                     hbaoRandomShort});
+
+    // TODO is it really dynamic?! i don't think so
+    state_->hbaoUniform.setData({nullptr, sizeof(HbaoUniformData)},
+                                Mn::GL::BufferUsage::DynamicDraw);
+  }
+
+  /* "init framebuffers" */
+  {
+    state_->sceneDepthLinear
+        .setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setStorage(1, Mn::GL::TextureFormat::R32F, configuration.size());
+
+    state_->depthLinear = Mn::GL::Framebuffer{{{}, configuration.size()}};
+    state_->depthLinear.attachTexture(Mn::GL::Framebuffer::ColorAttachment{0},
+                                      state_->sceneDepthLinear, 0);
+
+    state_->sceneViewNormal
+        .setMinificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setMagnificationFilter(Mn::GL::SamplerFilter::Nearest)
+        .setWrapping(Mn::GL::SamplerWrapping::ClampToEdge)
+        .setStorage(1, Mn::GL::TextureFormat::RGBA8, configuration.size());
+
+    state_->viewNormal = Mn::GL::Framebuffer{{{}, configuration.size()}};
+    state_->viewNormal.attachTexture(Mn::GL::Framebuffer::ColorAttachment{0},
+                                     state_->sceneViewNormal, 0);
+
+    const Mn::GL::TextureFormat aoFormat =
+        configuration.flags() & HbaoFlag::UseAoSpecialBlur
+            ? Mn::GL::TextureFormat::RG16F
+            : Mn::GL::TextureFormat::R8;
+    state_
+        ->hbaoResult
+        // TODO filter?!
```

Comment:
Obsolete TODOs, below also. I had marked these as I expected some issues, but there weren't any in the end.